clients/src/main/java/org/apache/kafka/common/record/LazyDownConvertingRecords.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.record;
+
+import org.apache.kafka.common.utils.Time;
+
+import java.io.IOException;
+import java.nio.channels.GatheringByteChannel;
+
+/**
+ * Records implementation which enables lazy down-conversion to avoid unnecessary memory utilization before
+ * a fetch response is ready to be sent.
+ */
+public class LazyDownConvertingRecords implements Records {
+    private final Records records;
+    private final long firstOffset;
+    private final byte toMagic;
+    private final Time time;
+    private ConvertedRecords<? extends Records> convertedRecords = null;
+
+    public LazyDownConvertingRecords(Records records,
+                                     byte toMagic,
+                                     long firstOffset,
+                                     Time time) {
+        this.records = records;
+        this.toMagic = toMagic;
+        this.firstOffset = firstOffset;
+        this.time = time;
+    }
+
+    @Override
+    public int sizeInBytes() {
+        return convertIfNeeded().records().sizeInBytes();
```

Comment:
`fetchResponseCallback` in `KafkaApis#handleFetchRequest` does:
```
brokerTopicStats.updateBytesOut(topicPartition.topic, fetchRequest.isFromFollower, data.records.sizeInBytes)
```
Won't that down-convert the batch, allocating memory while still on the I/O thread, before the batch is ready to be sent on the network thread?

Reply:
Ack. Good catch.